### PR TITLE
Get OutOfMemoryError to inherit from RuntimeError

### LINKED ISSUE #95936: Pylance: torch.cuda.OutOfMemoryError is not a valid exception class
### 🐛 Describe the bug

Pylance does not recognize torch.cuda.OutOfMemoryError as an exception. Using python extension on visual studio code.

Code runs properly, so this is just a typing issue.

## Example use case
```
try:
    self.model = self.model.to("gpu")
except torch.cuda.OutOfMemoryError: 
    self.model = self.model.to("cpu")
```

## Error
image

### Versions

PyTorch version: 1.13.1+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Home
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.9 (tags/v3.10.9:1dd9be6, Dec  6 2022, 20:01:21) [MSC v.1934 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.22621-SP0
Is CUDA available: True
CUDA runtime version: 12.0.76
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060 Ti
Nvidia driver version: 528.49
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture=9
CurrentClockSpeed=2100
DeviceID=CPU0
Family=198
L2CacheSize=2048
L2CacheSpeed=
Manufacturer=GenuineIntel
MaxClockSpeed=2100
Name=12th Gen Intel(R) Core(TM) i7-12700
ProcessorType=3
Revision=

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.2
[pip3] torch==1.13.1+cu116
[pip3] torchaudio==0.13.1+cu116
[pip3] torchvision==0.14.1+cu116
[conda] Could not collect

cc @ngimel @ezyang @malfet @rgommers @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by gerhean**
Get OutOfMemoryError to inherit from RuntimeError so that type checkers do not complain.
Also added a defined in comment to match with the other definitions.

Fixes #95936

### CODE REVIEW

File: torch/_C/__init__.pyi.in [hunk 1]
```diff
@@ -1839,7 +1839,8 @@ def _activate_cuda_trace() -> None: ...
 def _current_graph_task_id() -> _int: ...
 def _current_autograd_node() -> _Node: ...
 
-class _OutOfMemoryError: ...
+# Defined in torch/csrc/Exceptions.cpp
+class _OutOfMemoryError(RuntimeError): ...
```
Comment (Skylion007) [inline torch/_C/__init__.pyi.in:1843]: Rather annoying this doesn't inherit from MemoryError.

Comment (gerhean) [reply torch/_C/__init__.pyi.in:1843]: Would it be good to change the Exceptions.cpp bindings to PyExc_MemoryError then? Or would that possibly cause breaking changes? I'm new to Python C api so I'm not sure.

Comment (Skylion007) [reply torch/_C/__init__.pyi.in:1843]: Might be a breaking change, but one that will not affect a lot of projects in practice. This PR fixes the immediate typing issue though so perhaps we can merge this and then open a new PR that does try to see if we can move it to a PyExc_MemoryError without breaking everything.

**Review by Skylion007 — APPROVED** [review]

**Pull request closed without merging**